snuba/web/rpc/v1/endpoint_time_series.py

```diff
@@ -0,0 +1,203 @@
+    DataPoint,
+    TimeSeries,
+    TimeSeriesRequest,
+    TimeSeriesResponse,
+)
+
+from snuba.attribution.appid import AppID
+from snuba.attribution.attribution_info import AttributionInfo
+from snuba.datasets.entities.entity_key import EntityKey
+from snuba.datasets.entities.factory import get_entity
+from snuba.datasets.pluggable_dataset import PluggableDataset
+from snuba.query import OrderBy, OrderByDirection, SelectedExpression
+from snuba.query.data_source.simple import Entity
+from snuba.query.dsl import column
+from snuba.query.logical import Query
+from snuba.query.query_settings import HTTPQuerySettings
+from snuba.request import Request as SnubaRequest
+from snuba.web.query import run_query
+from snuba.web.rpc import RPCEndpoint
+from snuba.web.rpc.common.common import (
+    aggregation_to_expression,
+    attribute_key_to_expression,
+    base_conditions_and,
+    trace_item_filters_to_expression,
+    treeify_or_and_conditions,
+)
+from snuba.web.rpc.common.debug_info import (
+    extract_response_meta,
+    setup_trace_query_settings,
+)
+
+
+def _convert_result_timeseries(
+    request: TimeSeriesRequest, data: list[Dict[str, Any]]
+) -> Iterable[TimeSeries]:
+    # to convert the results, need to know which were the groupby columns and which ones
+    # were aggregations
+    aggregation_labels = set([agg.label for agg in request.aggregations])
+    group_by_labels = set([attr.name for attr in request.group_by])
+
+    # create a mapping with (all the group by attribute key,val pairs as strs, label name)
+    result_timeseries: dict[tuple[str, str], TimeSeries] = {}
+
+    # create a mapping for each timeseries of timestamp: row to fill data points not returned in the query
+    result_timeseries_timestamp_to_row: defaultdict[
+        tuple[str, str], dict[int, Dict[str, Any]]
+    ] = defaultdict(dict)
+    query_duration = (
+        request.meta.end_timestamp.seconds - request.meta.start_timestamp.seconds
+    )
+    time_buckets = [
+        Timestamp(seconds=(request.meta.start_timestamp.seconds) + secs)
+        for secs in range(0, query_duration, request.granularity_secs)
+    ]
+
+    for row in data:
+        group_by_map = {}
+
+        for col_name, col_value in row.items():
+            if col_name in group_by_labels:
+                group_by_map[col_name] = col_value
+
+        group_by_key = "|".join([f"{k},{v}" for k, v in group_by_map.items()])
+        for col_name in aggregation_labels:
+            if not result_timeseries.get((group_by_key, col_name), None):
+                result_timeseries[(group_by_key, col_name)] = TimeSeries(
+                    group_by_attributes=group_by_map,
+                    label=col_name,
+                    buckets=time_buckets,
+                )
+            result_timeseries_timestamp_to_row[(group_by_key, col_name)][
+                int(datetime.fromisoformat(row["time"]).timestamp())
+            ] = row
```

Comment:
what is group_by_map and group_by_key? Im not sure what you're doing here.

Reply:
I added an explanation, lemme know if it makes sense now